### PR TITLE
[processing] port some Python algorithms to C++

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Port some Python QGIS algorithms to C++:
- Aspect
- Hillshade
- Ruggedness index
- Slope
- Create constant raster
- Create attribute index
- Create spatial index
- Orthogonalize
- Create points layer from table
- Set M value
- Set Z value
- Single-sided buffer (distance parameter now dynamic)

## Checklist
- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmspatialindex.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/***************************************************************************
+                         qgsalgorithmspatialraster.cpp
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmspatialindex.cpp]:
```suggestion
                         qgsalgorithmspatialindex.cpp
```

**Review by nyalldawson — COMMENTED** [review]
Wow, mammoth effort -- great work!!

File: src/analysis/processing/qgsalgorithmconstantraster.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                QgsProcessingParameterNumber::Double, 0.1, false, 0.01 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "NUMBER" ), QObject::tr( "Constant value" ),
+                QgsProcessingParameterNumber::Double, 1, false ) );
+
+  addParameter( new QgsProcessingParameterRasterDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Constant" ) ) );
+}
+
+QVariantMap QgsConstantRasterAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsRectangle extent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context );
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmconstantraster.cpp]: Done

File: src/analysis/processing/qgsalgorithmpointslayerfromtable.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+{
+  return QStringLiteral( "vectorcreation" );
+}
+
+QString QgsPointsLayerFromTableAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm generates a points layer based on the values from an input table." )
+         + QStringLiteral( "\n\n" )
+         + QObject::tr( "The table must contain a field with the X coordinate of each point and another "
+                        "one with the Y coordinate. A CRS for the output layer has to be specified, and "
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmpointslayerfromtable.cpp]: Done

File: src/analysis/processing/qgsalgorithmconstantraster.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+QgsConstantRasterAlgorithm *QgsConstantRasterAlgorithm::createInstance() const
+{
+  return new QgsConstantRasterAlgorithm();
+}
+
+void QgsConstantRasterAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterExtent( QStringLiteral( "EXTENT" ), QObject::tr( "Desired extent" ) ) );
+  addParameter( new QgsProcessingParameterCrs( QStringLiteral( "TARGET_CRS" ), QObject::tr( "Target CRS" ), QStringLiteral( "ProjectCrs" ) ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "PIXEL_SIZE" ), QObject::tr( "Pixel size" ),
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmconstantraster.cpp]: Fixed

File: src/analysis/processing/qgsalgorithmspatialindex.cpp (hunk 1 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmspatialindex.cpp]: Fixed

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**